packages/compartment-mapper/src/hooks.md

```diff
@@ -0,0 +1,124 @@
+# Review of compartment-mapper hooks
+
+
+
+| Hook Name                 | Description                                                                           |
+|-------------------------- | --- |
+| `packageDataHook`         | Receives all found package descriptors data before graph translation.                                   |
+| `packageDependenciesHook` | Allows dynamic mutation of dependencies during node_modules graph translation.              |
+| `unknownCanonicalNameHook`| Called when the policy references unknown canonical names, can suggest typos/similar names. |
+| `moduleSourceHook`        | Invoked when a module source is created.                                             |
+| `packageConnectionsHook`  | Surfaces connections during digest. (ignored in archiving)                                               |
+
+
+
+[Type declarations for the hooks](./types/external.ts)
+
+```mermaid
+graph TB
+
+  exports((public exports))
+  exports -.- compartmentMapForNodeModules
+  exports -.- mapNodeModules
+  exports -.- loadLocation
+  exports -.- importLocation
+  exports -.- captureFromMap
+  exports -.- importFromMap
+  exports -.- loadFromMap
+  exports -.- digestCompartmentMap
+  exports -.- makeFunctor
+  exports -.- makeScript
+
+
+subgraph "import-hook.js"
+  moduleSourceHook{{moduleSourceHook}} -.- note5["for all module sources read"]
+  makeDeferError --> moduleSourceHook
+  makeImportHookMaker --> makeDeferError
+  makeImportNowHookMaker --> makeDeferError
+  makeImportNowHookMaker -- via importNowHook --> chooseModuleDescriptor --> executeLocalModuleSourceHook--> moduleSourceHook
+  makeImportHookMaker --via importHook exitModule logic--> moduleSourceHook
+  makeImportHookMaker --via importHook call--> chooseModuleDescriptor
+end
+
+subgraph "node-modules.js"
+  compartmentMapForNodeModules --> packageDataHook{{packageDataHook}} -.- note0["called once"]
+  compartmentMapForNodeModules --> unknownCanonicalNameHook{{unknownCanonicalNameHook}} -.- note1["for all issues from policy;<br>after defaultUnknownCanonicalNameHandler"]
+  compartmentMapForNodeModules --> translateGraph --> packageDependenciesHook{{packageDependenciesHook}} -.-note3["for all locatons in graph<br>after defaultPackageDependenciesFilter"]
+  mapNodeModules --> compartmentMapForNodeModules
+
+end
+
+subgraph "digest.js"
+  digestCompartmentMap --> translateCompartmentMap --> packageConnectionsHook{{packageConnectionsHook}} -.- note4["for all retained compartments"]
+end
+
+subgraph "bundle.js"
+  makeFunctor -- options:can include hooks ----------> mapNodeModules
+  makeScript -- options:can include hooks ----------> mapNodeModules
```

Comment:
(nit) Typesscript would complain, but it seems `MapNodeModulesOptions` would be successfully passed through these.